Add failing tests for exposing the value of .Self through an interface

danakj opened the pull request:
We test using .Self in a generic parameter of an interface and as the value of an associated constant.

danakj committed:
test-call-returning-facet-value

danakj commented:
I got a bad feeling I am getting the typish levels wrong here after continuing to stare at things.

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,55 @@ fn UseIndirect[T:! I](x: T) -> T {
+  //   - This finds `G` through the FacetType.
+  // - The second `.` is on a Call of type FacetAccessType into `BindSymbolicName` with type FacetType for `I`.
+  //   - This finds `G` through the FacetType (impl lookup strips off FacetAccessType).
+  // - The second `.` is on a Call of type FacetAccessType into `ImplWitnessAccess` of `I.X` into `LookupImplWitness`, which has type `type`
```
```suggestion
  // - The third `.` is on a Call of type FacetAccessType into `ImplWitnessAccess` of `I.X` into `LookupImplWitness`, which has type `type`
```

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,55 @@ fn UseIndirect[T:! I](x: T) -> T {
+  // - The second `.` is on a Call of type FacetAccessType into `BindSymbolicName` with type FacetType for `I`.
+  //   - This finds `G` through the FacetType (impl lookup strips off FacetAccessType).
+  // - The second `.` is on a Call of type FacetAccessType into `ImplWitnessAccess` of `I.X` into `LookupImplWitness`, which has type `type`
+  //   - Can't make calls on an `ImplWitnessAccess`.
```
FWIW, I would expect that the constant value of the `ImplWitnessAccess` would be the same type that we got for the second lookup.

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,55 @@ fn UseIndirect[T:! I](x: T) -> T {
+  // CHECK:STDERR:   U.G() as type;
+  // CHECK:STDERR:   ^~~~~~~~~~~~~
+  // CHECK:STDERR:
+  U.G() as type;
```
I think this one should fail: `U.G() as type` should be a value expression of type `U`, and we can't convert a value of type `U` to type `type`.

zygoloid replied:
Maybe `U.X as type` would be the thing we expect to work here. But that seems a little vacuous because it's already of type `type`.

danakj replied:
That sounds right, it's the FacetAccessType that is really messing with me.

Specifically, the return type here is `FacetAccessType(Special) of U` which knows U and knows the *type of U*. So then I want to treat the return as-if it's a value of type of the *type of U* which is a FacetType.

But that's wrong.. but I am having trouble finding where I fell off the path.

danakj replied:
Ok so I think I get it...

A facet value is a type (or a placeholder for a type) already, it just also holds a witness to say what interface(s) it's accessing.

So FacetAccessType isn't doing a type-of operation, it's just passing along the type that the facet value _is_. So the `FacetAccessType of U` is just `U` still, but asking for the final type that arrives in U.

danakj committed:
moretests

danakj commented:
Thanks for that review, I removed the bogus test, and I have added a bunch more tests of .Self that I was using to test #5966 while working on it. It would be good to make sure we agree they are all supposed to pass as written.

danakj committed:
add-file

danakj committed:
fail_todo_

danakj committed:
compounds

danakj committed:
adjust-nesting

danakj commented:
Updated the nesting tests with what we discovered in [open discussion today](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.0#heading=h.rg3417ebho7q).

danakj commented:
bump for review

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,99 @@ fn UseIndirect[T:! I](x: T) -> T {
+}
+
+fn F2[U:! I](T: U) {}
+fn F3[U:! I where .X = .Self](T: U) {}
```
```suggestion
fn F2[U:! I](v: U) {}
fn F3[U:! I where .X = .Self](v: U) {}
```
... so that it doesn't look like `T` is declaring a type.

danakj replied:
Ya, good point. Done

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,99 @@ fn UseIndirect[T:! I](x: T) -> T {
+  // CHECK:STDERR:   (U as type).G().G().G();
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  (U as type).G().G().G();
```
Maybe also add a test of `(U as type).G().G()`? From the underlining in the diagnostic it looks like that one would work too?

danakj replied:
Yep it does. Done.

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,99 @@ fn UseIndirect[T:! I](x: T) -> T {
+  // CHECK:STDERR:   F3(U.G().G().G());
+  // CHECK:STDERR:      ^~~~~~~~~~~
+  // CHECK:STDERR:
+  F3(U.G().G().G());
```
Do any of these work?
```
F2(U.G());
F2(U.G().G());
F3(U.G());
F3(U.G().G());
```

danakj replied:
Just F2(U.G()). No calls to F3 work atm. Added calls with a single .G() to each fn though.

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,99 @@ fn UseIndirect[T:! I](x: T) -> T {
+  // The returned value of `G` type `U` which has access to the methods of `I`.
+  //
+  // TODO: This works but a chained third call doesn't.
+  U.G().G();
```
It might be worth moving this out into its own split so that we can more easily ensure it doesn't regress.

danakj replied:
Done

zygoloid commented on toolchain/check/testdata/facet/period_self.carbon:
```diff
@@ -0,0 +1,498 @@
+// --- fail_period_self_as_type.carbon
+library "[[@TEST_NAME]]";
+
+interface I(T:! .Self) {
```
Presumably we should reject this binding even before we see uses of it. Worth adding a TODO?

danakj replied:
Yeah, done.

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,99 @@ fn UseIndirect[T:! I](x: T) -> T {
+
+fn F(U:! I where .X = .Self) {
+  // It's not possible to do compound member lookup here. You'd need
+  // to work with `U.(I.X)` instead.
```
Every time I read this test it takes me a minute to figure out what's going on, even with the reference to the issue. Maybe we can distill the explanation and include it directly here:

The second call looks for `v impls I`, not `U impls I`, where `v: U` is the non-type value returned by the first call.

Or maybe we could add another test before this one that makes it more obvious what's happening:
```carbon
fn F1(U:! I where .X = .Self) {
  // OK, type U converts to I
  let v: U = U.(I.G)();
  // error, non-type value v does not convert to I
  v.(I.G)();
}

fn F2(U:! I where .X = .Self) {
  // error, same as F1
  U.(I.G)().(I.G)();
}
```

danakj replied:
Ok I have reworked the test along these lines, thanks

zygoloid commented on toolchain/check/testdata/facet/period_self.carbon:
```diff
@@ -0,0 +1,498 @@
+fn G(U:! I(.Self) where .Self impls J(.Self)) {
+  // It's not possible to do compound lookup here. You would have to work with
+  // the original type `U` instead.
+  // See: https://github.com/carbon-language/carbon-lang/issues/6025
```
If you decide to make changes to the similar test in the other file, please also apply them here.

danakj replied:
Done

zygoloid commented on toolchain/check/testdata/facet/period_self.carbon:
```diff
@@ -0,0 +1,498 @@
+  let a: U = U.G();
+}
+
+// --- fail_template_access.carbon
```
I think this one should be a `fail_todo`, but I'm also not sure if we want to be testing `template` stuff here. Seems like that's out of scope (this doesn't even involve `.Self`). Maybe either delete this test or move it to `testdata/generic/template`?

danakj replied:
Right, moved it.

zygoloid commented on toolchain/check/testdata/facet/period_self.carbon:
```diff
@@ -0,0 +1,498 @@
+  // CHECK:STDERR:   U as (I(T) where .A ={} and .B = {});
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  U as (I(T) where .A ={} and .B = {});
```
```suggestion
  U as (I(T) where .A = {} and .B = {});
```

danakj replied:
Done

zygoloid commented on toolchain/check/testdata/facet/period_self.carbon:
```diff
@@ -0,0 +1,498 @@
+
+// TODO: This should be an error: The third `.Self` becomes undefined. It's not
+// referring to `T` but we need each explicit `.Self` to be replaced with
+// something at the end of the expression.
```
Is this an error because `.Self` is undefined or because it's ambiguous? I would expect the latter: it could name `T` or, for (eg) `U:! T.A`, it could name `U`.

danakj replied:
Well, maybe it's a matter of perspective. I would prefer that we don't define .Self at all if it _would be_ ambiguous, or would be left unbound (at the end of the statement?). But the root reason is the inability to bind to something unambiguous. Reworded.

danakj committed:
review

danakj committed:
review

danakj reviewed (commented):
PTAL

danakj committed:
typo

danakj committed:
Merge branch 'trunk' into test-call-returning-facet-value

danakj committed:
diagnostics-changed

danakj commented on toolchain/check/testdata/facet/period_self.carbon:
```diff
@@ -0,0 +1,517 @@
+  // CHECK:STDERR:
+  let u: T = U.G();
+
+  // Shows both `I(.Self)` are `I(T)`.
+  // CHECK:STDERR: fail_todo_nested_period_self.carbon:[[@LINE+4]]:9: error: found cycle in facet type constraint for `.(I(T).A)` [FacetTypeConstraintCycle]
+  // CHECK:STDERR:   T as (I(T) where .A = (I(T) where .A = {} and .B = {}));
+  // CHECK:STDERR:         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  T as (I(T) where .A = (I(T) where .A = {} and .B = {}));
```
I don't think this should actually be a cycle, but it's still being reported as one when I get .Self working more, so pointing this out to double check.

zygoloid replied:
Yeah, I agree this is a false positive. I guess our cycle checking is incorrectly thinking the `.A` nested inside the second `where` is a rewrite for the outer `.A`, or something like that?

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,190 @@ fn UseIndirect[T:! I](x: T) -> T {
+  // CHECK:STDERR:   U as (I where .I1 = U);
+  // CHECK:STDERR:   ^~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  U as (I where .I1 = U);
```
Do we have tests of the opposite condition somewhere too?
```
U as (I where .I1 = .Self);
(U as type) as (I where .I1 = .Self);
```

zygoloid commented on toolchain/check/testdata/facet/access.carbon:
```diff
@@ -80,6 +80,190 @@ fn UseIndirect[T:! I](x: T) -> T {
+  // CHECK:STDERR:   u.(I.G)();
+  // CHECK:STDERR:   ^~~~~~~
+  // CHECK:STDERR:
+  u.(I.G)();
```
Would it make sense to split this and the next line out from the file that contains the previous let declaration, so that we don't have a mixture of "should not fail but does" and "should fail and does" in the same file?

zygoloid commented on toolchain/check/testdata/generic/template/fail_todo_template_access_assoc_const.carbon:
```diff
@@ -0,0 +1,23 @@
+library "[[@TEST_NAME]]";
+
+// TODO: Improved error: no member named `I1` in `T` of type `type`.
+//
```
I think this ought to be valid: because `T` is a template parameter, we ought to defer the lookup until we have a non-template value for it.

zygoloid reviewed (approved):
Looks good, thanks. Some comments but nothing that needs another look I think.

danakj committed:
review

danakj committed:
Merge remote-tracking branch 'origin/trunk' into test-call-returning-facet-value

danakj committed:
apply-core-destroy

danakj committed:
test-type-and-self-parameter
